Library/Homebrew/software_spec.rb

```diff
@@ -162,12 +162,27 @@ def depends_on(spec)
     add_dep_option(dep) if dep
   end
 
-  def uses_from_macos(spec, _bounds = {})
-    spec = [spec.dup.shift].to_h if spec.is_a?(Hash)
+  def uses_from_macos(deps, bounds = {})
+    if deps.is_a?(Hash)
+      bounds = deps.dup
+      deps = [bounds.shift].to_h
+    end
+
+    @uses_from_macos_elements << deps
 
-    @uses_from_macos_elements << spec
+    if OS.mac?
```

Comment:
I'm not sure I understand why we're moving `OS.mac?` checks away from `extend/os`? Apologies that this is probably due to my feedback.

Reply:
No worries! Let me know if this answers your question. If not, I think I can make a simple example to help.

At some point when we start running a Homebrew command, we `require "formula"` which in turn runs [`require "software_spec"`](https://github.com/Homebrew/brew/blob/985f29f5951e37456b2215cf17f80741fa273860/Library/Homebrew/formula.rb#L16). This defines a [`uses_from_macos` method](https://github.com/Homebrew/brew/blob/985f29f5951e37456b2215cf17f80741fa273860/Library/Homebrew/software_spec.rb#L165-L171) that always passes its arguments on to `depends_on` (this is the Linux implementation of `uses_from_macos`.

Then, at the bottom of this file, we call [`require "extend/os/software_spec"`](https://github.com/Homebrew/brew/blob/985f29f5951e37456b2215cf17f80741fa273860/Library/Homebrew/software_spec.rb#L603) which executes [this file](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/software_spec.rb) and, depending on the system configuration _at that moment_, requires either the macOS or Linux extension files:
- If it requires the macOS extension, the `uses_from_macos` method is [re-defined](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/software_spec.rb) as the method that checks the `since` condition to see whether to pass to `depends_on` or not.
- If it requires the [Linux extension](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/linux/software_spec.rb), there are no changes to `uses_from_macos` (so it retains its original functionality).

The problem arises when we want to simulate Linux while on macOS or vice versa. If we switch from one to the other, we don't actually re-execute any of the files listed above because their contents have already been executed and their modules/methods have already been defined. So, we are stuck with the implementation for the _original_ system configuration (i.e. whatever it was when we first ran `require "formula"`. So, if we simulate Linux on macOS and re-load a formula, we'll still execute the Linux version of `uses_from_macos` (i.e. always add the dependency) which is incorrect for macOS.

Similarly, the `ignore_missing_libraries` method is an immediate error by default, and is re-defined without the error on Linux. So, if we start on macOS and then switch to Linux, we still execute the macOS version of `ignore_missing_libraries` which throws an error and halts execution.

Reply:
Thanks, that makes sense. What I don't understand is why `OS.mac?` is queried here rather than a `SimulateSystem` method?

Reply:
Ah, because with the newest version `OS.mac?` reflects the simulation as well as the reality (just to try it out). Based on discussion above I think we'd rather go the other route where `OS.mac?` is only when we're actually on a mac and `SimulateSystem.simulating_or_running_on_linux?` is preferable

Reply:
Ah, yes. Agreed on changing based on the above and if this is changed to a `SimulateSystem` check: that's fine with me 👍🏻

Reply:
If we're already not distinguishing between simulating or not, it seems a bit superfluous to distinguish between `OS.linux?` and `SimulateSystem.simulating_or_running_on_linux?`, though.

It's not a super strong opinion, but I feel it'll be simpler to write future code if all we cared about is `OS.foo?` instead of considering also that we might want a code path to be run while simulating so we also need `SimulateSystem.simulating_or_running_on_linux?`.

Reply:
> If we're already not distinguishing between simulating or not

I was planning on reverting this change so that `OS.mac?` etc only reflect the actual system and is not affected by simulations. Does that remove your concern?

Reply:
> but I feel it'll be simpler to write future code if all we cared about is `OS.foo?`

That's my point, though: we don't want to do this. I've worked pretty hard to avoid scattering `OS.mac?` and `OS.linux?` all over the codebase in favour of the (IMO much more maintainable) `extend/os` model.

As @Rylan12 has addressed above: we have to do this a bit differently with simulation given that we need to do this without e.g. `require`.

Reply:
Okay, I've pushed some changes, let me know if this is what you both were expecting.

Now, `OS.mac?` is only true when we're actually on macOS. `Homebrew::SimulateSystem.simulating_or_running_on_macos?` is true if we're actually on macOS or are simulating macOS.